core/promise/consumer.go

```diff
@@ -36,17 +66,49 @@ func (c *Consumer) NewRequest() (requestPtr interface{}) {
 
 // Consume handles requests from endpoint and replies with response
 func (c *Consumer) Consume(requestPtr interface{}) (response interface{}, err error) {
-	// request := requestPtr.(*Request)
+	request, ok := requestPtr.(*Request)
+	if !ok {
+		return nil, errUnsupportedRequest
+	}
 
-	// TODO there should be some validation of the received proposal and storing it somewhere for the server needs.
-	// TODO signature validation of the promise should be here too.
+	receivedPromise, err := json.Marshal(request.SignedPromise.Promise)
+	if err != nil {
+		return nil, err
+	}
 
-	// if request.SignedPromise.IssuerSignature == valid {
-	// 		return &Response{
-	// 			Success: false,
-	// 			Message: fmt.Sprintf("Bas signature: %s", request.SignedPromise.IssuerSignature),
-	// 		}, fmt.Errorf("Bad promise signature")
-	// }
+	signature := identity.SignatureBase64(string(request.SignedPromise.IssuerSignature))
+	issuer := identity.FromAddress(request.SignedPromise.Promise.IssuerID)
+	verifier := identity.NewVerifierIdentity(issuer)
+	if !verifier.Verify(receivedPromise, signature) {
+		return failedResponse(errBadSignature, request), nil
+	}
+
+	benefiter := identity.FromAddress(request.SignedPromise.Promise.BenefiterID)
+	if benefiter.Address != c.proposal.ProviderID {
+		return failedResponse(errUnknownBenefiter, request), nil
+	}
+
+	price := c.proposal.PaymentMethod.GetPrice()
+	amount := request.SignedPromise.Promise.Amount.Amount
+	if amount < price.Amount {
+		return failedResponse(errLowAmount, request), nil
+	}
+
+	balance, err := c.etherClient.BalanceAt(context.Background(), common.HexToAddress(issuer.Address), nil)
+	if err != nil {
+		return nil, err
+	}
+	if balance.Uint64() < amount {
+		return failedResponse(errLowBalance, request), nil
+	}
+
+	if err := c.storage.Store(issuer.Address, &request.SignedPromise.Promise); err != nil {
+		return nil, err
+	}
+
+	return &Response{Success: true, Message: "Promise accepted", Request: request}, nil
+}
 
-	return &Response{Success: true, Message: "Promise accepted"}, nil
+func failedResponse(err error, request *Request) *Response {
+	return &Response{Success: false, Message: err.Error(), Request: request}
```

Comment:
We dont need to leak text errors. Actually it should be system-only error codes (see `session/create_dto.go`)

Reply:
I believe that we need to send error details to the client. 

Errors like `Internal error` from the server says that something happened internally on the server and user should retry request later changing nothing. 

But if we are sending him a message like `your promise format invalid` or `you have too low balance` it says that user did something wrong and he needs to change it before next sending. 

It will be very hard to debug client problem without propper error responses if user uses external exit node. 

I think it's good practice to responce with as much details as we can.

Reply:
We already had these discussions and made decisions:
- protocol should not too verbose with errors
- it can be attack vector
- we not building app here, but crystallising protocol messages
- there might be other nodes, which wont respond with real error codes, but will fake them. So you dont control infra and you cant depend on error codes anyway

Reply:
Keep in mind, that in protocol we are coding for standardised error-codes, instead of text messages.
Because if it's text messages, which language should it be? English, Esperanto? :)

https://github.com/mysteriumnetwork/node/blob/master/session/create_dto.go